File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/ast/empty_catch_block.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""
Forbid empty catch block rule.
"""

import sys


from collections.abc import Iterable

import libcst as cst

from rules.harness_check_rule import (
    Finding,
    FindingFix,
    FixSafety,
    HarnessCheckRule,
    parse_python,
)
from core.rule_context import RuleContext, relative, stack_sources_configured

sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")


class EmptyCatchBlockRule(HarnessCheckRule):
    """Validate emptyCatchBlock check."""

    category = "emptyCatchBlock"

    def applies(self, ctx: RuleContext) -> bool:
        """Check if this rule applies to the context."""
        if not ctx.is_enabled(self.category):
            return False
        return stack_sources_configured(ctx.manifest.raw, self.category)

    def validate(self, ctx: RuleContext) -> Iterable[Finding]:
        """Validate emptyCatchBlock check."""
        severity = ctx.severity_of(self.category)
        category = self.category
        sources = ctx.stack_sources(self.category)

        class EmptyCatchFinder(cst.CSTVisitor):
            METADATA_DEPENDENCIES = (cst.metadata.PositionProvider,)

            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path

            def visit_Try(self, node: cst.Try) -> bool:
                for handler in node.handlers:
                    if not isinstance(handler.body, cst.IndentedBlock):
                        continue
                    body_stmts = handler.body.body
                    if not body_stmts:
                        pos = self.get_metadata(cst.metadata.PositionProvider, handler)
                        self.findings.append(
                            Finding(
                                severity,
                                category,
                                f"{self.rel_path}:{pos.start.line}: empty catch block; handle, rethrow, or convert to a Finding",
                                file=self.rel_path,
                                start_line=pos.start.line,
                                start_column=pos.start.column + 1,
                                end_line=pos.end.line,
                                end_column=pos.end.column + 1,
                                fix=FindingFix(
                                    description="add exception handling or rethrow",
                                    safety=FixSafety.UNSAFE,
                                    edits=(),
                                ),
                            )
                        )
                return True

        def collect_findings():
            for path in sources:
                tree, error = parse_python(path)
                if error is not None:
                    yield Finding(
                        severity,
                        category,
                        f"{relative(path, ctx.root)}: syntax error: {error}",
                    )
                    continue
                wrapper = cst.MetadataWrapper(tree)
                visitor = EmptyCatchFinder(relative(path, ctx.root))
                wrapper.visit(visitor)
                yield from visitor.findings
        return list(collect_findings())


RULE: HarnessCheckRule = EmptyCatchBlockRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/ast/empty-catch-block.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type { Node, SourceFile } from "typescript@6.0.3";
import {
  createSourceFile,
  isCatchClause,
  SyntaxKind,
} from "typescript@6.0.3";
import type {
  Finding,
  HarnessCheckRule,
  RuleContext,
} from "../harness-check-rule";
import { astChildrenOf } from "../../core/ast-traversal";

/**
 * Forbid empty catch blocks.
 */
export const emptyCatchBlockRule: HarnessCheckRule = {
  category: "emptyCatchBlock",
  applies(ctx: RuleContext): boolean {
    return true;
  },

  validate(ctx: RuleContext): readonly Finding[] {
    return ctx.stackSources("emptyCatchBlock").flatMap((file) => {
      const text = ctx.read(file);
      if (!text) {
        return [];
      }

      const sourceFile: SourceFile = createSourceFile(
        file,
        text,
        SyntaxKind.LatestVersion,
        true,
      );

      const visitNode = (node: Node): readonly Finding[] => {
        if (!isCatchClause(node) || node.block.statements.length !== 0) {
          return astChildrenOf(node).flatMap(visitNode);
        }
        const start = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
        const end = sourceFile.getLineAndCharacterOfPosition(node.getEnd());
        return [{
          severity: ctx.severityOf("emptyCatchBlock"),
          category: "emptyCatchBlock",
          message: `${file}:${start.line + 1}: empty catch block; handle, rethrow, or convert to a Finding`,
          file,
          startLine: start.line + 1,
          startColumn: start.character + 1,
          endLine: end.line + 1,
          endColumn: end.character + 1,
          fix: {
            description: "add catch block handling or rethrow statement",
            safety: "unsafe",
            edits: [],
          },
        }, ...astChildrenOf(node).flatMap(visitNode)];
      };

      return visitNode(sourceFile);
    });
  },
};
